File: rustserver/src/handlers/positioncount.rs
use actix_web::{web, HttpResponse};
use crate::db;
use mongodb::bson::doc;

pub async fn get_position_count(path: web::Path<(String, String)>, q: web::Query<std::collections::HashMap<String, String>>, client: web::Data<mongodb::Client>) -> HttpResponse {
    let (coinName, tableName) = path.into_inner();
    let side = q.get("side").map(|s| s.as_str());

    if let Some(s) = side {
        if s != "Long" && s != "Short" {
            return HttpResponse::BadRequest().json(serde_json::json!({"error":"Invalid side. Must be \"Long\" or \"Short\""}));
        }
    }

    let mut filter = doc!{"coinName": {"$regex": format!("^{}$", coinName), "$options": "i"}, "status": "open" };
    if let Some(s) = side { filter.insert("positionSide", s); }

    let coll = db::db(&client).collection::<mongodb::bson::Document>(&tableName);
    match coll.count_documents(filter, None).await {
        Ok(count) => HttpResponse::Ok().json(serde_json::json!({"message":"Position count retrieved successfully","coinName": coinName, "tableName": tableName, "status":"open","side": side.unwrap_or("all".into()), "count": count})),
        Err(e) => HttpResponse::InternalServerError().json(serde_json::json!({"error": e.to_string()})),
    }
}

File: api/routes/positioncount.js
const router = require('express').Router();
const { getCollection } = require('../utils/database');
router.get("/version",(req,res)=>res.send("Version 1.0"))
// GET /getPositionCount/:coinName/:tableName?side=Long|Short
// Returns the total count of positions for a specific coin with status 'open'
// Optional: ?side=Long or ?side=Short to filter by position side
router.get('/getPositionCount/:coinName/:tableName', async (req, res) => {
  try {
    const { coinName, tableName } = req.params;
    const { side } = req.query;

    // Validate side parameter if provided
    if (side && !['Long', 'Short'].includes(side)) {
      return res.status(400).json({ error: 'Invalid side. Must be "Long" or "Short"' });
    }

    // Build query filter
    let filter = {
      coinName: { $regex: `^${coinName}$`, $options: 'i' },
      status: 'open'
    };

    if (side) {
      filter.positionSide = side;
    }

    const collection = getCollection(tableName);
    const count = await collection.countDocuments(filter);

    res.json({
      message: 'Position count retrieved successfully',
      coinName,
      tableName,
      status: 'open',
      side: side || 'all',
      count
    });
  } catch (err) {
    console.error('Error fetching position count:', err);
    return res.status(500).json({ error: 'Error fetching position count' });
  }
});

module.exports = router;
